src/Stratis.Bitcoin.Features.Miner/PosMinting.cs

```diff
@@ -447,17 +453,29 @@ public void StopStake()
 
             while (!this.stakeCancellationTokenSource.Token.IsCancellationRequested)
             {
-                // Wait until we have at least one connected peer who's headers are synced with ours AND until not in IBD.
-                while (!this.connection.ConnectedPeers.Any(x => x.Behavior<ChainHeadersBehavior>().IsSynced())
-                    || this.initialBlockDownloadState.IsInitialBlockDownload())
+                bool readyToMine = !this.initialBlockDownloadState.IsInitialBlockDownload();
+
+                // If not in IBD- check if fully synced.
+                if (readyToMine)
                 {
-                    if (this.initialBlockDownloadState.IsInitialBlockDownload()) this.logger.LogTrace("Waiting for IBD to complete...");
-                    else this.logger.LogTrace("Waiting to be connected with at least one synced network peer...");
+                    // Most advanced peer's pending tip.
+                    ChainedBlock chainedBlock = this.connection.ConnectedPeers
```

Comment:
@noescape00 Can we maybe write this better?

I see that Behaviors is an extension of NetworkPeerCollection... How about we move that whole linq statement into NetworkPeerCollection? Then you can just go:

`this.connection.ConnectedPeers.HighestPendingTip;`

That way you can write a quick unit test for it? In other words we are moving calculated logic where it belongs and making it testable 💯

Reply:
@dangershony already convinced me to avoid using BlockRepository here so this one is no longer applicable 🙂